src/to-have-accessibility-state.ts

```diff
@@ -0,0 +1,73 @@
+import type { AccessibilityState } from 'react-native';
+import type { ReactTestInstance } from 'react-test-renderer';
+import { matcherHint, stringify } from 'jest-matcher-utils';
+import { checkReactElement, getMessage, stringifyDefined } from './utils';
+
+export function toHaveAccessibilityState(
+  this: jest.MatcherContext,
+  element: ReactTestInstance,
+  expectedState: AccessibilityState,
+) {
+  checkReactElement(element, toHaveAccessibilityState, this);
+
+  const impliedState = getAccessibilityState(element);
+  return {
+    pass: matchAccessibilityState(element, expectedState),
+    message: () => {
+      const matcher = matcherHint(
+        `${this.isNot ? '.not' : ''}.toHaveAccessibilityState`,
+        'element',
+        stringify(expectedState),
+      );
+      return getMessage(
+        matcher,
+        `Expected the element ${this.isNot ? 'not to' : 'to'} have acessibility state`,
+        stringify(expectedState),
+        'Received element with implied accessibility state',
+        stringifyDefined(impliedState),
+      );
+    },
+  };
+}
+
+export const accessibilityStateKeys: (keyof AccessibilityState)[] = [
+  'disabled',
+  'selected',
+  'checked',
+  'busy',
+  'expanded',
+];
+
+/**
+ * Default accessibility state values based on experiments using accessibility
+ * inspector/screen reader on iOS and Android.
+ *
+ * @see https://github.com/callstack/react-native-testing-library/wiki/Accessibility:-State
+ */
+const defaultState: AccessibilityState = {
+  disabled: false,
+  selected: false,
+  checked: undefined,
+  busy: false,
+  expanded: undefined,
+};
```

Comment:
This behaviour is not driven by RN but rather by native side of both iOS and Android. E.g. on iOS `disable` state corresponds to `notEnabled` a11y trait. So if you omit disabled state or put explicit `disabled: false` that will result in not having `notEnabled` trait (yeah duble negation). Similar for `selected` (iOS `selected` trait) and `busy`. 

This is different form `checked` and `expanded` that have a different behaviour when being explicitly unchecked than not having related checked/mixed/unchecked at all. I've done an empirical research on iOS and ANdroid using a11y inspectors/screen readers to figure that out: https://github.com/callstack/react-native-testing-library/wiki/Accessibility:-State

Regarding possibility of change, while it's possible I find it unlikely since the behaviour is largely based on native OS behaviour, so RN itself does not have a lot of free hand here.

Reply:
Alright got it!